crypto/dh/dh_key.c

```diff
@@ -125,6 +125,18 @@ static int generate_key(DH *dh)
             l = dh->length ? dh->length : BN_num_bits(dh->p) - 1;
             if (!BN_priv_rand(priv_key, l, BN_RAND_TOP_ONE, BN_RAND_BOTTOM_ANY))
                 goto err;
+            /*
+             * handle known cases where g is a quadratic non-residue:
+             * for g = 2: p % 8 == 3
+             * for g = 5: p % 5 != 4
+             */
+            if ((BN_is_word(dh->g, DH_GENERATOR_2) && !BN_is_bit_set(dh->p, 2))
+                || (BN_is_word(dh->g, DH_GENERATOR_5)
+                    && BN_mod_word(dh->p, 5) != 4)) {
+                /* clear bit 0, since it won't be a secret anyway */
+                if (!BN_clear_bit(priv_key, 0))
+                    goto err;
+            }
```

Comment:
Probably harmless, but is this useful or necessary?

Reply:
Certainly harmless.
One might argue, it saves one additional modular multiplication by g.

Reply:
In the mod 5 case, what is the cost of computing `p mod 5` every time we generate a key?  Is it worth it?  Since `256 mod 5` is one, the residue is just the sum of all the octets of `p` mod 5, which can perhaps be computed faster than a generic mod 5 reduction.  But do we really case about `g = 5`?  Does anybody actually use that?

Reply:
no, I'd bet 5, is never used.  and it is probably more expensive than what it could save theoretically.